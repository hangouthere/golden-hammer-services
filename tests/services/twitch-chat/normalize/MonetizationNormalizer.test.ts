const normalizerClass = require('-/twitch-chat/normalize/MonetizationNormalizer');

const MOCK_DATA = {
  userState: {
    'tmi-sent-ts': '1649705365314'
  }
};
describe('Twitch Chat: Normalizer - Monetization', () => {
  let normalizer;

  beforeEach(() => {
    normalizer = new normalizerClass();
  });

  describe('Value Estimation', () => {
    it('should throw an error on invalid sub plan types', () => {
      expect(() => normalizer._estimateValue('invalidType', NaN)).toThrow();
    });

    it('should NOT throw an error on valid sub plan types', () => {
      expect(() => normalizer._estimateValue('Prime', 1)).not.toThrow();
    });

    it('should calculate plan types according to Twitch Tiers', () => {
      expect(normalizer._estimateValue('Prime', 2)).toBe(4.99);
      expect(normalizer._estimateValue('1000', 2)).toBe(4.99);
      expect(normalizer._estimateValue('2000', 2)).toBe(7.992000000000001);
      expect(normalizer._estimateValue('3000', 2)).toBe(14.993999999999998);
    });
  });

  describe('Normalizing', () => {
    beforeEach(() => {
      normalizer._estimateValue = vitest.fn().mockReturnValue(4.2);
    });

    it('should normalize a Cheer', () => {
      const normalized = normalizer.normalizeCheer({
        incomingEventArguments: [
          {
            ...MOCK_DATA.userState,
            'display-name': 'userCheer',
            bits: 5000
          },
          'Cheer Message'
        ]
      });

      expect(normalized).toMatchSnapshot();
    });

    it('should normalize a Subscription', () => {
      const normalized = normalizer.normalizeSubscription({
        incomingEventArguments: [
          'userSubscribed',
          { prime: true },
          'Sub Message',
          {
            ...MOCK_DATA.userState,
            'display-name': 'userCheer',
            'msg-param-sub-plan': 'Prime'
          }
        ]
      });

      expect(normalized).toMatchSnapshot();
    });

    it('should normalize a Resub', () => {
      const normalized = normalizer.normalizeResub({
        incomingEventArguments: [
          'userReSubscribed',
          null,
          'ReSub message',
          {
            ...MOCK_DATA.userState,
            'display-name': 'userResub',
            'msg-param-sub-plan': 'Prime'
          },
          { prime: true }
        ]
      });

      expect(normalized).toMatchSnapshot();
    });

    it('should normalize a SubGift Send', () => {
      const normalized = normalizer.normalizeSubgiftSend({
        incomingEventArguments: [
          'userSendingGift',
          1,
          null,
          {
            ...MOCK_DATA.userState,
            'system-msg': 'Sub Gift Msg',
            'msg-param-sub-plan': '2000'
          }
        ]
      });

      expect(normalized).toMatchSnapshot();
    });

    it('should normalize a SubGift Recieve', () => {
      const normalized = normalizer.normalizeSubgiftRecieve({
        incomingEventArguments: [
          'userSendingGift',
          null,
          'userRecievingGift',
          { prime: true },
          {
            ...MOCK_DATA.userState,
            'system-msg': 'Sub Recieve Msg',
            'msg-param-sub-plan': '2000'
          }
        ]
      });

      expect(normalized).toMatchSnapshot();
    });

    describe('Durations from input', () => {
      it('should test for msg-param-months', () => {
        let normalized = normalizer.normalizeSubscription({
          incomingEventArguments: [
            'userSubscribed',
            { prime: false },
            'Sub Message',
            {
              ...MOCK_DATA.userState,
              'display-name': 'userSub',
              'msg-param-sub-plan': '2000',
              'msg-param-months': 4
            }
          ]
        });

        expect(normalized).toMatchSnapshot();

        normalized = normalizer.normalizeResub({
          incomingEventArguments: [
            'userReSubscribed',
            null,
            'ReSub message',
            {
              ...MOCK_DATA.userState,
              'display-name': 'userResub',
              'msg-param-sub-plan': '2000',
              'msg-param-months': 4
            },
            { prime: false }
          ]
        });

        expect(normalized).toMatchSnapshot();

        normalized = normalizer.normalizeSubgiftRecieve({
          incomingEventArguments: [
            'userSendingGift',
            null,
            'userRecievingGift',
            { prime: false },
            {
              ...MOCK_DATA.userState,
              'system-msg': 'Sub Recieve Msg',
              'msg-param-sub-plan': '2000',
              'msg-param-months': 4
            }
          ]
        });
      });

      it('should fallback to 1 if no msg-param-months', () => {
        let normalized = normalizer.normalizeSubscription({
          incomingEventArguments: [
            'userSubscribed',
            { prime: false },
            'Sub Message',
            {
              ...MOCK_DATA.userState,
              'display-name': 'userSub',
              'msg-param-sub-plan': '2000'
            }
          ]
        });

        expect(normalized).toMatchSnapshot();

        normalized = normalizer.normalizeResub({
          incomingEventArguments: [
            'userReSubscribed',
            null,
            'ReSub message',
            {
              ...MOCK_DATA.userState,
              'display-name': 'userResub',
              'msg-param-sub-plan': '2000'
            },
            { prime: false }
          ]
        });

        expect(normalized).toMatchSnapshot();

        normalized = normalizer.normalizeSubgiftRecieve({
          incomingEventArguments: [
            'userSendingGift',
            null,
            'userRecievingGift',
            { prime: false },
            {
              ...MOCK_DATA.userState,
              'system-msg': 'Sub Recieve Msg',
              'msg-param-sub-plan': '2000'
            }
          ]
        });
      });
    });
  });
});
