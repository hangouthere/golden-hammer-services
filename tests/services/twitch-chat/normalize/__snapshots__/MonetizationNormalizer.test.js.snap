// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Twitch Chat: Normalizer - Monetization Normalizing Durations from input should fallback to 1 if no msg-param-months 1`] = `
Object {
  "normalizedData": Object {
    "duration": 1,
    "estimatedValue": 4.2,
    "message": "Sub Message",
    "sourceUserName": "userSubscribed",
  },
  "timestamp": 1649705365314,
}
`;

exports[`Twitch Chat: Normalizer - Monetization Normalizing Durations from input should fallback to 1 if no msg-param-months 2`] = `
Object {
  "normalizedData": Object {
    "duration": 1,
    "estimatedValue": 4.2,
    "message": "ReSub message",
    "sourceUserName": "userReSubscribed",
  },
  "timestamp": 1649705365314,
}
`;

exports[`Twitch Chat: Normalizer - Monetization Normalizing Durations from input should test for msg-param-months 1`] = `
Object {
  "normalizedData": Object {
    "duration": 4,
    "estimatedValue": 4.2,
    "message": "Sub Message",
    "sourceUserName": "userSubscribed",
  },
  "timestamp": 1649705365314,
}
`;

exports[`Twitch Chat: Normalizer - Monetization Normalizing Durations from input should test for msg-param-months 2`] = `
Object {
  "normalizedData": Object {
    "duration": 4,
    "estimatedValue": 4.2,
    "message": "ReSub message",
    "sourceUserName": "userReSubscribed",
  },
  "timestamp": 1649705365314,
}
`;

exports[`Twitch Chat: Normalizer - Monetization Normalizing should normalize a Cheer 1`] = `
Object {
  "normalizedData": Object {
    "estimatedValue": 50,
    "message": "Cheer Message",
    "sourceUserName": "userCheer",
  },
  "timestamp": 1649705365314,
}
`;

exports[`Twitch Chat: Normalizer - Monetization Normalizing should normalize a Resub 1`] = `
Object {
  "normalizedData": Object {
    "duration": 1,
    "estimatedValue": 4.2,
    "message": "ReSub message",
    "sourceUserName": "userReSubscribed",
  },
  "timestamp": 1649705365314,
}
`;

exports[`Twitch Chat: Normalizer - Monetization Normalizing should normalize a SubGift Recieve 1`] = `
Object {
  "normalizedData": Object {
    "duration": 1,
    "estimatedValue": 4.2,
    "message": "Sub Recieve Msg",
    "sourceUserName": "userSendingGift",
    "targetUserName": "userRecievingGift",
  },
  "timestamp": 1649705365314,
}
`;

exports[`Twitch Chat: Normalizer - Monetization Normalizing should normalize a SubGift Send 1`] = `
Object {
  "normalizedData": Object {
    "estimatedValue": 4.2,
    "message": "Sub Gift Msg",
    "sourceUserName": "userSendingGift",
  },
  "timestamp": 1649705365314,
}
`;

exports[`Twitch Chat: Normalizer - Monetization Normalizing should normalize a Subscription 1`] = `
Object {
  "normalizedData": Object {
    "duration": 1,
    "estimatedValue": 4.2,
    "message": "Sub Message",
    "sourceUserName": "userSubscribed",
  },
  "timestamp": 1649705365314,
}
`;
